Replace aquasecurity/libbpfgo with cilium/ebpf - BPF

jakule opened the pull request:
Closes https://github.com/gravitational/teleport/issues/36099
Closes https://github.com/gravitational/teleport.e/issues/2997

jakule committed:
Fixes after rebase

Remove debug statements and unused code in BPF library

This commit eliminates unnecessary debug log statements from bpf_test.go and removes unused code in disk.go and network.go files within BPF library. The aim is to make codebase cleaner, enhancing readability and maintainability without losing track of functional aspects. It's part of an ongoing effort to maintain optimal code quality across the project.

Revise Dockerfile and Makefile instructions

Adjusted Dockerfile and Makefile commands to optimize build process. The changes ensure better readability and maintainability of code, especially with respect to BPF bytecode and test preparations. A clause was also added to check for the presence of clang before building BPF bytecode.

Refactor BPF code and tests

Refactored BPF source code and its corresponding tests for better readability and efficiency. This includes updating data structures like rawExecEvent, rawOpenEvent, rawConn4Event and rawConn6Event to their more meritorious versions like commandDataT, diskDataT, networkIpv4DataT and networkIpv6DataT respectively. Also, generated concise and clear bindigs for eBPF programs to replace the previous multi-arch setup. Furthermore, correspondingly adapted all related test cases arising in bpf_test.go, command.go and other files.

Remove counter_test from BPF and related test code

This commit removes the file counter_test.bpf.c from the enhanced recording in BPF. Following this file deletion, corresponding test functions in bpf_test.go, that were originally testing the functionality provided by counter_test.bpf.c, have been removed as well.

Refactor BPF test and improve error handling

The commit refactors the BPF test suite to run each test as a sub-test by using t.Run, easing debugging and error tracing. It also simplifies the 'executeCommand' function by removing redundant function calls related to finding executable files, and improves error handling. In 'cgroup.go', the code has been adjusted to retry on EBUSY errors when removing a cgroup, instead of ignoring them.

Update cgroup deletion process and error handling

The changes made introduce a retry mechanism when deleting a cgroup in case the cgroup is busy. This prevents a failed attempt from immediately returning an error. Additionally, error handling has been improved in the 'writePids' function to better handle 'no such process' errors.

Enhance cgroup monitoring in disk.bpf.c

The code has been updated to check if current cgroup is being monitored in the "enter_open" method of "disk.bpf.c". If the cgroup is not marked for monitoring, it will be ignored, enhancing overall cgroup management.

Improve resource handling

This commit introduces a deferred close to the sendEvents function to ensure proper resource handling of the eventBuf object.

Add rlimit package to remove resource limits

Introduced the "github.com/cilium/ebpf/rlimit" package and added a function to remove resource limits for kernels less than 5.11. This operation is necessary for kernels with a lower version because they may face issues when running applications with allocated memory limitations.

Fix a test after rebase

Refactor BPF testing setup and implement session closing

This commit updates the BPF testing structure to remove the context dependency in event generation functions, simplifying the test setup. Additionally, it adds a 'closed' field to the 'open', 'conn', and 'exec' structs alongside mutex protection. This implementation allows checking if a session is closed before any operations, correctly handling already closed sessions. Furthermore, it corrects tracepoints for handling 'openat' and 'openat2' system calls for different platforms. This refactoring effort aims to improve the robustness and readability of the BPF testing suite.

Update conditions for BPF bytecode building

This update modifies the Makefile to add a conditional check for the presence of /usr/include/bpf/bpf_helpers.h. This new requirement must be met along with the existence of /usr/include/linux/bpf.h and the installation of Clang, in order to facilitate the correct build of BPF bytecode. Consequently, this change enhances the reliability of the build process.

Add conditional for bpf_helpers.h in Makefile

This commit adds an additional check in the Makefile for the existence of /usr/include/bpf/bpf_helpers.h. This ensures that BPF bytecode only gets built when both bpf.h and bpf_helpers.h exist, and Clang is installed, improving the build process reliability.

Refactor and improve BPF bytecode generation

This commit refactors and improves the BPF bytecode generation process. It introduces distinct bytecode structures for different architectures, ensuring better compatibility and execution. The BPF test is adjusted due to its tendency to fail unpredictably, hence it's commented out to avoid unnecessary build failures. This commit also eliminates certain redundancies in `common.mk`, improving the clarity of the build process.

Comment out test code in bpf_test.go

The test code in the file bpf_test.go has been commented out. These changes are circumstantial and do not imply that the test code is obsolete or unnecessary. The use of such tests may be needed in future commits or for local development testing and debugging.

Remove libbpf dependency from Vagrantfile and update README

The Vagrantfile and README have been updated to remove the no longer needed libbpf dependency. This change simplifies the project set up by eliminating unused build steps in the Vagrantfile and correcting related instructions in the README. This results in a more streamlined, intuitive build process and reduced potential for code complexity and dependency conflicts.

Remove obsolete dependencies and related code blocks

The 'aquasecurity/libbpfgo' library referencing and the related code blocks have been removed from the test files and core modules. The corresponding setup instructions in the dockerfiles and build scripts have also been eliminated. This update simplifies the build process and overall project structure while reducing unnecessary dependency risks.

Remove aquasecurity/libbpfgo dependency

Remove restricted.bpf.c

Migrate disk events to cilium

Migrate network events to cilium

Remove restricted session

Use cilium/ebpf for BPF

zmb3 commented on lib/bpf/disk_bpfel.o:
oops!

jakule replied:
@zmb3 Not oops. This is my plan. I don't see any reason why we need to generate those files on every build. In order to generate the BPF bytecode, you still need `libbpf` installed, as it comes with some BPF C headers and Clang 11+. eBPF files, once generated, don't need to be recompiled with every build.
That's a similar approach to what we do with proto files, and IMO, having generated files in the repo is easier to live with (this will let us remove clang and libbpf from our ci/release images).

jakule commented on lib/bpf/disk.go:
```diff
@@ -74,10 +78,13 @@ type cgroupRegister interface {
+
+	eventBuf chan []byte
+	toClose  []io.Closer
+	//lost     *Counter
```
@russjones Do you know if the BPF counters are used by anyone? From what I see, they are defined and initialized but not used. I want to remove them (including tests), but I'm curious if maybe you're aware of any history behind it. Is there any reason not to remove them?

reedloden commented on build.assets/Dockerfile-bpf:
```diff
@@ -0,0 +1,26 @@
+# Minmal container fo generating BPF bytecode
```
```suggestion
# syntax=docker/dockerfile:1

# Minimal container for generating BPF bytecode
```

jakule replied:
Do we need this header even if I'm not using any "extended" docker syntax?

reedloden replied:
Nope, but just for consistency among all dockerfiles.

But there is a typo fix as well -- s/Minmal/Minimal/

reedloden commented on build.assets/Makefile:
```diff
@@ -276,6 +274,18 @@ fix-imports: buildbox
+.PHONY: bpf-bytecode
+bpf-bytecode:
+	docker build \
+  		--build-arg UID=$(UID) \
```
```suggestion
        --build-arg UID=$(UID) \
```

reedloden reviewed (commented):
Woot woot. :-)

jakule committed:
Update github.com/cilium/ebpf to v0.16.0

This commit updates the `github.com/cilium/ebpf` dependency from version `v0.12.3` to `v0.16.0` in the `go.mod` file. This change ensures compatibility with the latest features and bug fixes provided in the newer version.

jakule committed:
Fix ut

jakule committed:
Terraform go mod tidy
